### PR TITLE
Alerting: Central alert history (part1)

### PR DESCRIPTION
**Pull request by soniaAguilarPeiron**
**What is this feature?**

This PR is the part1 for the implementation of the new Central Alert State History.
For it, this PR uses scenes for creating this new view. This allow us rendering the bar chart panel and also sharing a common time period interval. The main scene is the `CentralAlertHistoryScene` , which is responsible of rendering the time period picker , plus the two nested scenes:

  1. Scene for rendering the `events bar chart`.
  2. Scene for rendering the `list of events`. This scene is a custom scene object (`HistoryEventsListObjectRenderer`). We need to render this component  in a scene to be able to share the common time period filter with the bar chart. As scenes can share parent time ranges. See [scene docs ](https://github.com/grafana/scenes/blob/main/docusaurus/docs/advanced-custom-scene-objects.md)

⚠️This feature is still behind the `alertingCentralAlertHistory` feature toggle

**For running this feature locally:** 

- You need to [enable Loki for the alert state history ](https://grafana.com/docs/grafana/latest/alerting/set-up/configure-alert-state-history/)

- rename the Loki data source you want to use for history with `grafanacloud-alert-state-history`

- enabling the FF  `alertingCentralAlertHistory`  in your browser with :
`window.localStorage.setItem('grafana.featureToggles', 'alertingCentralAlertHistory=true')`

I particular in this PR, these are the implemented tasks:

- [x] Add the feature toggle
- [x] Add the search and time period inputs
- [x] Add the bar chart for the number of events.
- [x] Add the table for the events with rows collapsed

**Remaining tasks (not solved in this PR):**

- [ ] Implement the expanded rows detail.
- [ ] Implement events chart using a runtime data source (using api instead of requesting directly the Loki data source)
- [ ] Share labels filter (it's not straight forward as labels in Loki are not the labels we get from the line field, that are the ones we use in the list). 
- [ ] Add tests
- [ ] Protect feature behind a new permission (menu and route), and depending of the annotations implementation.
- [ ] RBAC has to be defined. We are going to use a new permission. Once we use runtime data source, bar chart will be using the same RBAC.
- [ ] Show transitions cause (when present in the state) in a way it's more obvious at a glance(not only a tooltip)
- [ ] Use tooltip for transitions instead of state
- [ ] review correct format date depending on the time zone?
- [ ] Paginate events list (as Virtualized list only performs well in less than 10k items, and in this case, it's easy to exceed that number due to the numerous transitions some customers are dealing with.)

[Link to the issue](https://github.com/grafana/grafana/issues/88345)

**Why do we need this feature?**

In this PR we solve the tasks mentioned in the issue except the last one (expanding the rows)

**Who is this feature for?**

All users.

**Special notes for your reviewer:**

https://github.com/grafana/grafana/assets/33540275/fb824648-59ce-4590-8b94-282fb1b7ea3a

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    </Stack>
+  );
+}
+
+function EventState({ state }: { state: GrafanaAlertStateWithReason }) {
+  const styles = useStyles2(getStyles);
+  if (!isGrafanaAlertState(state)) {
+    // missing series , not sure what to render here. They are always normal state?
+    return (
+      <Tooltip content={'Missing series'}>
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:233]: I think in this case it's important to know the cause of the transition

File: public/app/features/alerting/routes.tsx [hunk 2]
```diff
@@ -169,6 +169,19 @@ export function getAlertingRoutes(cfg = config): RouteDescriptor[] {
         () => import(/* webpackChunkName: "AlertGroups" */ 'app/features/alerting/unified/AlertGroups')
       ),
     },
+    {
+      path: '/alerting/history/',
+      roles: evaluateAccess([
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/routes.tsx:174]: we are going to use a new permission. For now, as the feature is behind a ff and it's in progress, we keep this access control.

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+  const { historyRecords } = useRuleHistoryRecords(stateHistory, eventsFilter);
+
+  if (isError) {
+    return <HistoryErrorMessage error={error} />;
+  }
+
+  return (
+    <Stack direction="column" gap={1}>
+      <Stack gap={1} direction="row" justifyContent={'space-between'}>
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:85]: I don't think we need this Stack here since it only has a single child element. Are we planning to add something here later?

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+interface HistoryLogEventsProps {
+  logRecords: LogRecord[];
+}
+function HistoryLogEvents({ logRecords }: HistoryLogEventsProps) {
+  // display log records
+  return (
+    <ul>
+      {logRecords.map((record, index) => {
+        return <EventRow key={index} record={record} />;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:117]: I don't think this is a good key – when refresh is turned on each item might receive a new index and every row will be re-rendered. We should figure out a better way to identify a log line – I think `timestamp + hash(labels)` should be unique enough.

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+interface EventTransitionProps {
+  previous: GrafanaAlertStateWithReason;
+  current: GrafanaAlertStateWithReason;
+}
+function EventTransition({ previous, current }: EventTransitionProps) {
+  return (
+    <Stack gap={0.5} direction={'row'}>
+      <EventState state={previous} />
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:227]: These are probably not final, the design in the Figma file is a tad outdated – [this file](https://github.com/grafana/grafana/blob/main/public/app/features/alerting/unified/components/rule-list/RuleListIcon.tsx) has the updated icons for state / health (but doesn't include missing series).

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+  const styles = useStyles2(getStyles);
+
+  if (!isGrafanaAlertState(state) && !isAlertStateWithReason(state)) {
+    return (
+      <Tooltip content={'No recognized state'}>
+        <Icon name="exclamation-triangle" size="md" />
+      </Tooltip>
+    );
+  }
+  const stateWithReason = mapStateWithReasonToBaseState(state);
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:244]: nit: this is actually the `stateWithoutReason` :)

Perhaps to simplify things we should write a single function that extract both the "base" state and the "reason" from the alert state. We're currently using two different approaches to do that.

Probably something as simple as

```typescript
type State = GrafanaAlertState | PromAlertingRuleState

function parseState (state: string): [State, string | undefined] {
  const match = state.match(/(?<state>.*) \((?<reason>.*)\)/);
  if (!match?.groups?.state) {
    throw new Error('failed to parse state');
  }
  
  const parsedState = match.groups?.state as State;
  const parsedReason = match.groups?.reason;
  
  return [parsedState, parsedReason];
}
```

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+  } = useGetRuleHistoryQuery(
+    {
+      from: from,
+      to: to,
+      limit: LIMIT_EVENTS,
+    },
+    {
+      refetchOnFocus: true,
+      refetchOnReconnect: true,
+      pollingInterval: STATE_HISTORY_POLLING_INTERVAL,
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:73]: Would this conflict with the refresh interval from the scenes?

File: public/app/features/alerting/unified/components/rules/state-history/common.ts [hunk 8]
```diff
@@ -7,6 +7,7 @@ export interface Line {
   current: GrafanaAlertStateWithReason;
   values?: Record<string, number>;
   labels?: Record<string, string>;
+  ruleUID?: string;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/state-history/common.ts:10]: 🙌

**Review by gillesdemey — COMMENTED** [review]
Great work! 🥳 

I think we can also convert some of the existing visuals in the existing alert rule state history in the future – so best to keep that in mind when designing the component interfaces. There's already a lot of good stuff in here that we should be able to leverage there too to achieve consistency.

File: public/locales/en-US/grafana.json [hunk 9]
```diff
@@ -84,15 +94,15 @@
     },
     "counts": {
       "alertRule_one": "{{count}} alert rule",
-      "alertRule_other": "{{count}} alert rules",
+      "alertRule_other": "{{count}} alert rule",
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/locales/en-US/grafana.json:97]: yes, I don't know why 🤔. But looking to the rest of` *_others` are not plural, so I guess it's correct

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 6 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:244]: As the `mapStateWithReasonToBaseState` was already implemented I found adding the new function for getting only the base state more reasonable. I will update the naming for the function and for the variable though.

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+import { STATE_HISTORY_POLLING_INTERVAL } from '../state-history/LokiStateHistory';
+import { LogRecord } from '../state-history/common';
+import { useRuleHistoryRecords } from '../state-history/useRuleHistoryRecords';
+
+const LIMIT_LABELS = 5;
+const LIMIT_EVENTS = 250;
+
+const HistoryEventsList = ({ timeRange }: { timeRange?: TimeRange }) => {
+  const styles = useStyles2(getStyles);
+  const { useGetRuleHistoryQuery } = stateHistoryApi;
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:48]: I was using the same notation we use in [the existing code in alert state history](https://github.com/grafana/grafana/blob/8d36949f61c0158753029f6fb84c7548b2572b46/public/app/features/alerting/unified/components/rules/state-history/LokiStateHistory.tsx#L32) , is there any reason why you prefer this notation?

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 3 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:86]: Oops, this stack might be from temporary code I used while implementing this component.

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 5 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:228]: That could be great!, but I'm not sure if I can use this component , as in the Loki line we have `GrafanaAlertStateWithReason`, while in the RuleListIcon we have `PromAlertingRuleState`. Do you still think we can? 🤔
```
export enum PromAlertingRuleState {
  Firing = 'firing',
  Inactive = 'inactive',
  Pending = 'pending',
}

export enum GrafanaAlertState {
  Normal = 'Normal',
  Alerting = 'Alerting',
  Pending = 'Pending',
  NoData = 'NoData',
  Error = 'Error',
}

type GrafanaAlertStateReason = ` (${string})` | '';

export type GrafanaAlertStateWithReason = `${GrafanaAlertState}${GrafanaAlertStateReason}`;
```

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:48]: Yep, using the `endpoints` notation it will correctly (at least in my editor) allow me to jump to the definition of that endpoint, I can't seem to make that work with de-structuring the hooks. Maybe it's a local setup problem though.

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 5 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:228]: updated the icons!

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:48]: good point! 💯

File: public/app/features/alerting/unified/components/AlertLabels.tsx [hunk 11]
```diff
@@ -28,6 +29,7 @@ export const AlertLabels = ({ labels, commonLabels = {}, size }: Props) => {
 
   const commonLabelsCount = Object.keys(commonLabels).length;
   const hasCommonLabels = commonLabelsCount > 0;
+  const tooltip = t('alert-labels.button.show.tooltip', 'Show common labels');
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/AlertLabels.tsx:32]: changes in this file are because are not related to the pr, because was a change i reverted (adding a limit to these props). And betterer forced me to add translations.

File: public/app/features/alerting/unified/components/AlertLabels.tsx [hunk 12]
```diff
@@ -14,11 +15,13 @@ interface Props {
   labels: Record<string, string>;
   commonLabels?: Record<string, string>;
   size?: LabelSize;
+  limit?: number;
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/AlertLabels.tsx]: I reverted adding the line. And I decided to make the labels col more flexible. The alert name also has now a tooltip for seeing the whole name.

https://github.com/grafana/grafana/assets/33540275/7aac5a80-dc7e-4b1a-a94e-cc9428eff9dd

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 7 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:73]: good point, let's split it

**Comment by gillesdemey** [comment]
I'm not sure if we can work around this, but I've noticed that using the refresh interval from the scenes the entire list is briefly empty re-rendered as opposed to using the regular 10 second interval, that one seems to incrementally add items to the list just fine 🤔

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+  const { register, handleSubmit, reset } = useForm({ defaultValues: { query: '' } }); //  form for search field
+  const from = timeRange?.from.unix();
+  const to = timeRange?.to.unix();
+  const onFilterCleared = useCallback(() => {
+    setEventsFilter('');
+    reset();
+  }, [setEventsFilter, reset]);
+
+  const {
+    currentData: stateHistory,
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:60]: I believe we should be using `data` here instead of `currentData`, that way the list view won't flash empty (temporarily) while we fetch new data after the timerange has been updated by the scene.

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+  } = stateHistoryApi.endpoints.getRuleHistory.useQuery(
+    {
+      from: from,
+      to: to,
+      limit: LIMIT_EVENTS,
+    },
+    {
+      refetchOnFocus: true,
+      refetchOnReconnect: true,
+      pollingInterval: STATE_HISTORY_POLLING_INTERVAL,
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:73]: ohh, good point!!

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistoryScene.tsx [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+import { DataSourceInformation, PANEL_STYLES } from '../../../home/Insights';
+import { SectionSubheader } from '../../../insights/SectionSubheader';
+
+import { HistoryEventsListObjectRenderer } from './CentralAlertHistory';
+
+export const CentralAlertHistoryScene = () => {
+  const dataSourceSrv = getDataSourceSrv();
+  const ashDs: DataSourceInformation = {
+    type: 'loki',
+    uid: 'grafanacloud-alert-state-history',
```
Comment (tomratcliffe) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistoryScene.tsx:35]: (Just noting the things that didn't work locally for me without changes, as per Slack)
I had to rename this to match my already-set-up-Loki (`gdev-loki`)

File: public/locales/en-US/grafana.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomratcliffe) [reply public/locales/en-US/grafana.json:95]: This doesn't seem right though 🤔 Something seems off as they wouldn't now make sense, but I don't know enough about the translations without diving into it

**Review by tomratcliffe — COMMENTED** [review]
Looks good overall, functionality looks smart 😎 

I had some nit changes that were quicker for me to make in a branch rather than suggest for you, so I put them in another PR to this one: https://github.com/grafana/grafana/pull/89138

Feel free to bring them in, or that can go into main afterwards if you'd rather just get this in first

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+  return <Alert title={title}>{stringifyErrorLike(error)}</Alert>;
+}
+
+interface SearchFieldInputProps {
+  showClearFilterSuffix: boolean;
+  onClearFilterClick: () => void;
+}
+const SearchFieldInput = React.forwardRef<HTMLInputElement, SearchFieldInputProps>(
+  ({ showClearFilterSuffix, onClearFilterClick, ...rest }: SearchFieldInputProps, ref) => {
+    const placeholder = t('central-alert-history.filter.placeholder', 'Filter events in the list by labels');
```
Comment (jtheory) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:138]: Phrasing: this would be either:

* "Filter events in the list by label" (singular "label", even though they might use more than one)
* or "Filter events in the list with labels"

The first is a little more idiomatic & natural-sounding to my ear. Maybe the second is better for people who don't have English as a first language.

File: public/locales/en-US/grafana.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/locales/en-US/grafana.json:95]: We are not using this key, so, I believe it's something they changed in main 🤷‍♀️

[the remark this answers is not in the document]
Comment (tomratcliffe) [reply public/locales/en-US/grafana.json:95]: It's this, I believe:

https://www.i18next.com/translation-function/plurals#singular-plural

Relating to how plurals are handled

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+  }
+}
+
+interface TimestampProps {
+  time: number; // epoch timestamp
+}
+
+const Timestamp = ({ time }: TimestampProps) => {
+  const dateTime = new Date(time);
+  const formattedDate = dateTime.toLocaleString('en-US', {
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:292]: adding this question in todo. Thanks!

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistoryScene.tsx (hunk 15 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistoryScene.tsx:35]: I added some notes in the PR description with the steps to run this PR locally

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 16 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:138]: Thanks Rob! I will update the text with your suggestion. With the final PR we will ping Brenda for reviewing all the texts in the feature. Not pinging her yet , as these might change

File: public/locales/en-US/grafana.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/locales/en-US/grafana.json:95]: I mean the parent key

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

**Comment by soniaAguilarPeiron** [comment]
> Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.
> 
> Thank you!


no problem! thanks @diegommm

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+    isError,
+    error,
+  } = stateHistoryApi.endpoints.getRuleHistory.useQuery(
+    {
+      from: from,
+      to: to,
+      limit: LIMIT_EVENTS,
+    },
+    {
+      refetchOnFocus: true,
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:70]: nit: These would also not be really useful as the data will never change even if the window is re-focussed or the network is re-connected – the stale data should still be valid :)

**Review by gillesdemey — APPROVED** [review]
LGTM – on to part 2! 🚀

**Review by tomratcliffe — APPROVED** [review]
(I'm also happy, looks great!)

File: public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx (hunk 18 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/rules/central-state-history/CentralAlertHistory.tsx:70]: 💯
